Enterprise-CMCS/macpro-onemac
# Create infrastructure for systemwide and user notifications in OneMAC Application

Story: https://qmacbis.atlassian.net/browse//OY2-30181
Endpoint: https://d3o6kg6uaumkqe.cloudfront.net/

### Details
> As a developer, I want to create a notification banner infrastructure that can deliver a set of messages to the front end. These messages could be either system messages displayed for all users or user specific messages that are displayed per user. The per user messages should be able to be removed through a dismiss operation.

### Changes
**backend:** 
created the following api endpoints:
`getActiveSystemNotifications`: will get all active system notifications based on publication and expiry date
`getActiveUserNotifications`: will get all active user notifications that have been published since last user login
`dismissUserNotification`: will mark a user notification as dismissed
`createUserNotifications`: create new records for the given user that are active and not yet in the users pk; returns the collection of notifications created; this method will also dismiss any previous notifications the user had as active

**frontend:**
`NotificationApi`: API for all things Notifications
`NotificationBanner`: component for the user Notification's that appear at the top of every page
`NotificationCard`: component for system wide Notifications which appear on the home page 

### Implementation Notes
**backend:**
-  getActiveUserNotifications turned out to not be utilized since we combined the functionality into the createUserNotifications as it already returns the notifications that were created. Combine that with the fact that this method is only called upon login (or when not in local storage) and for this implementation notifications should only be displayed on first login, therefore only new notifications need to be displayed. We decided to keep the function since it was already built and could be utilized in the future.
**frontend:**
- using context in App.tsx to store user notifications so that it displays across all the application
- context does not store on refresh, so for user notifications we also used local storage 
- the props for the notification components use the same type as returned from notificationApi 
- the system notification has an API call on the home.js page

https://qmacbis.atlassian.net/wiki/spaces/MACPRO/pages/3657957377/User+Notification+System+Architecture+and+Implementation

### Screenshots 

![Screenshot 2024-09-23 at 12 24 34 PM](https://github.com/user-attachments/assets/8eb1aea3-12b3-4d1d-b73a-cb2f4db87b19)

### Test Plan

1. Insert test records to DB (both system and user_notification)
2. Verify system notification is displayed on home page even when unauthenticated
3. Verify upon login the user notification is displayed for the user and the system notification is still displayed on home page
4. Navigate within application and verify user notification continues to display
5. Refresh or close browser and reload and verify user notification still displays
6. Dismiss user notification and navigate application; verify user notification no longer appears
7. Logout and the Login and verify user notification is no longer there (Retry this step also without first manually dismissing)
